### DISCUSSION TITLE
Toasts and redirect in mount() method

### DISCUSSION THREAD

**Discussion by pbourdet** [livewire/livewire#9095] [1. Help]
### Livewire version

v3.5.17

### Laravel version

v11.35.1

### Which PHP version are you using?

PHP 8.4

### What is the problem?

I have a full page component with a `Gate`, that redirects the user if they are not allowed to see the page. Before redirecting, there is a toast added using `Flux::toast` method.

But the toast does not show up on the page. I'm guessing this is because event dispatched are maybe not kept on redirection ? I've found a somewhat hacky solution but was wondering if something built-in exists.

Thanks !

### Code snippets

Component : 

```PHP
class MyFullPageComponent extends Component
{
    public function mount(): void
    {
        if (!Gate::allows('view', MyModel::class)) {
            Flux::toast('You cannot.', variant: 'danger');
            $this->redirectRoute('dashboard', navigate: true);
        }
    }
}
```

I have this in my `base.blade.php` : 

```
@persist('toast')
    <flux:toast position="top right"/>
@endpersist
```

My solution for now is to use the following : 

```PHP
//MyFullPageComponent
session()->flash('toast', ['text' => 'You cannot.', 'variant' => 'danger']);

//base.blade.php
@if (null !== session('toast'))
    <script>
        document.addEventListener('DOMContentLoaded', function() {
            Flux.toast({text: '{{ session('toast')['text'] }}', variant: '{{ session('toast')['variant'] }}'});
        });
    </script>
@endif
```

### Wirebox URL

No wirebox 

### How do you expect it to work?

I was expecting the toast to show after redirecting

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my help request will be removed if I haven't met the criteria above.

**Comment by Aleksandar-Mitic** [comment]
This is also problem for me. It would be really nice if we get some general solution for this.

**Comment by snits35m** [comment]
Same here! I also use session variables for now - but there should be a better solution!

**Comment by pravnkay** [comment]
Yeah, I was expecting it to work naturally. But it doesn't.

**Comment by ghabriel25** [comment]
See here https://laracasts.com/discuss/channels/livewire/toast-question?page=1&replyId=975890 for the example of implementation

**Comment by yudin-s** [comment]
If the redirect happens from `mount()`, the component often never gets a chance to render a browser event/toast on the current page.

For the simple case I would use session flash and display the toast on the destination page:

```php
session()->flash('toast', [
    'type' => 'success',
    'message' => 'Saved successfully.',
]);

return redirect()->route('dashboard');
```

Then the destination layout/component reads the flash value and triggers the toast after hydration.

One caveat from lazy page components: if something consumes the flash before the lazy component hydrates, the message can disappear. In that case use a normal session `put()`/`pull()` pattern, or explicitly `reflash()` during the lifecycle where the destination still needs the message.

For example, conceptually:

```php
session()->put('toast', [...]);

// destination component
$toast = session()->pull('toast');
```

If you really need to show the toast before redirecting, make it a two-step client interaction: dispatch the toast, wait briefly or wait for user acknowledgement, then redirect. For most admin flows, though, flash/session-on-destination is more reliable than dispatching a browser event right before navigation.

**Reply by ghabriel25** [reply]
@yudin-s Need to mention that flashed session doesnt survive on lazy page component. For lazy page component, it needs session to be put-and-pulled or reflash the session from `dehydrate()` in lazy page component
